builtin/game/register.lua

```diff
@@ -295,8 +207,120 @@ function core.register_tool(name, tooldef)
 			end
 		end
 	end
+end
+
+local default_tables = {
+	node = core.nodedef_default,
+	craft = core.craftitemdef_default,
+	tool = core.tooldef_default,
+	none = core.noneitemdef_default,
+}
+
+local preprocess_fns = {
+	node = preprocess_node,
+	craft = preprocess_craft,
+	tool = preprocess_tool,
+}
+
+function core.register_item(name, itemdef)
+	-- Check name
+	if name == nil then
+		error("Unable to register item: Name is nil")
+	end
+	name = check_modname_prefix(tostring(name))
+	if forbidden_item_names[name] then
+		error("Unable to register item: Name is forbidden: " .. name)
+	end
+
+	itemdef.name = name
+
+	-- Compatibility stuff depending on type
+	local fn = preprocess_fns[itemdef.type]
+	if fn then
+		fn(itemdef)
+	end
+
+	-- Apply defaults
+	local defaults = default_tables[itemdef.type]
+	if defaults == nil then
+		error("Unable to register item: Type is invalid: " .. dump(itemdef))
+	end
+	local old_mt = getmetatable(itemdef)
+	-- TODO most of these checks should become an error after a while (maybe in 2026?)
+	if old_mt ~= nil and next(old_mt) ~= nil then
+		-- Note that even registering multiple identical items with the same table
+		-- is not allowed, due to the 'name' property.
+		if old_mt.__index == defaults then
+			core.log("warning", "Item definition table was reused between registrations. "..
+				"This is unsupported and broken: " .. name)
+		else
+			core.log("warning", "Item definition has a metatable, this is "..
+				"unsupported and it will be overwritten: " .. name)
+		end
+	end
+	setmetatable(itemdef, {__index = defaults})
 
-	core.register_item(name, tooldef)
+	-- BEGIN Legacy stuff
+	if itemdef.cookresult_itemstring ~= nil and itemdef.cookresult_itemstring ~= "" then
+		core.register_craft({
+			type="cooking",
+			output=itemdef.cookresult_itemstring,
+			recipe=itemdef.name,
+			cooktime=itemdef.furnace_cooktime
+		})
+	end
+	if itemdef.furnace_burntime ~= nil and itemdef.furnace_burntime >= 0 then
+		core.register_craft({
+			type="fuel",
+			recipe=itemdef.name,
+			burntime=itemdef.furnace_burntime
+		})
+	end
+	-- END Legacy stuff
+
+	itemdef.mod_origin = core.get_current_modname() or "??"
+
+	-- Ignore new keys as a failsafe to prevent mistakes
+	getmetatable(itemdef).__newindex = function() end
+
+	-- Add to registered_* tables
+	if itemdef.type == "node" then
```

Comment:
:shrug: btw could also use a table `{node = core.registered_nodes, craft = core.registered_craftitems, tool = core.registered_tools}` here (and could reuse that table in the deletion code to delete from the right table)

(but it doesn't matter)